File: src/MrHotel/ApiService/Rooms/Data/RoomUpdateData.cs
namespace MrHotel.ApiService.Rooms.Data;

using MrHotel.ApiService.RoomPropertyGroups.Services;
using MrHotel.Database.Entities.Rooms;

public record RoomUpdateData(
    string Name,
    IReadOnlyCollection<Guid> PropertiesIds)
{
    public async Task<RoomUpdate> ToRoomUpdate(
        RoomPropertyGroupManager propertyGroupManager)
    {
        IEnumerable<RoomPropertyGroup> groups = await propertyGroupManager.GetPropertyGroups();
        IEnumerable<RoomProperty> properties = this.GetIncludedProperties(groups);

        return new RoomUpdate(this.Name, properties.ToList());
    }

    private IEnumerable<RoomProperty> GetIncludedProperties(IEnumerable<RoomPropertyGroup> groups)
    {
        return groups
            .SelectMany(g => g.Properties)
            .Where(p => this.PropertiesIds.Contains(p.Id));
    }
}

File: src/MrHotel/WebApp/src/app/services/room-manager/data/room-update-data.ts
import { Guid, Stringify } from "@customTypes/index";

export class RoomUpdateData {
  private readonly _name: string;
  private readonly _propertiesIds: Guid[];

  public constructor(name: string, propertiesIds: Guid[]) {
    this._name = name;
    this._propertiesIds = propertiesIds;
  }

  public toJSON(): Stringify<RoomUpdateData> {
    return {
      name: this._name,
      propertiesIds: this._propertiesIds,
    };
  }
}
